Components: migrate `ConfirmDialog` component's Stories from knobs to controls

chad1008 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Migrate the Stories for the `ConfirmDialog` component from knobs to the more modern controls API. The one exception is the `Controlled` story, which benefits more from the current format's code snippet, so doesn't have any active Controls.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
See https://github.com/WordPress/gutenberg/issues/35665 for more details, but due to knobs being deprecated, all components will eventually need to be migrated.

## Testing Instructions
1. `npm run storybook:dev`
2. Confirm that Knobs are no longer available, and Controls have taken their place
3. Confirm the DocsPage looks correct with no errors. The data will be somewhat incomplete, but this should be addressed in a future PR when the stories are migrated to TypeScript.

chad1008 committed:
ConfirmDialog Stories: implement a template, migrate 'Default' and 'With Custom Button Labels' from knobs to controls

chad1008 committed:
ConfirmDialog Stories: migrate 'Very Long Message' story from knobs to controls

chad1008 committed:
ConfirmDialog Stories: migrate 'With JSX Message' story from knobs to controls

chad1008 committed:
ConfirmDialog Stories: migrate 'Uncontrolled with explicit cancel' story from knobs to controls

chad1008 committed:
ConfirmDialog: add Story demonstrating 'Title' prop

chad1008 committed:
ConfirmDialog: Update 'Controlled' story description

chad1008 committed:
ConfirmDialog Stories: add argTypes data

chad1008 committed:
ConfirmDialog: convert 'With JSX Message' story to provide actual JSX examples

chad1008 committed:
ConfirmDialog: use default args for stories where appropriate

chad1008 committed:
Revert "ConfirmDialog: add Story demonstrating 'Title' prop"

This reverts commit 04961e7df536307e769cd6bb20e0fb8274d8ddb5.

chad1008 committed:
ConfirmDialog Stories: clean up story descriptions

chad1008 committed:
ConfirmDialog Stories: clean up argTypes data

mirka commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -15,97 +15,121 @@ import Button from '../../button';
 import { ConfirmDialog } from '..';
 
-export default {
+const daText = () =>
+	text( 'message', 'Would you like to privately publish the post now?' );
```
Could we eliminate this knob?

chad1008 replied:
Removed! ✅

mirka commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -15,97 +15,121 @@ import Button from '../../button';
-			<Heading level={ 1 }>{ confirmVal }</Heading>
-		</>
-	);
+// JSX elements can be passed as children to further customize the dialog.
+export const WithJSXMessage = Template.bind( {} );
+WithJSXMessage.args = {
+	jsxChildren: '<Heading level={ 2 }>A JSX Heading</Heading>',
+};
+WithJSXMessage.parameters = {
+	controls: { exclude: 'children' },
```
I think we could remove this story, since people just assume that `children` takes arbitrary elements unless noted otherwise.

chad1008 replied:
Excellent. I didn't love the options for demonstrating JSX strings in a control all that much anyway, so not having to is a win 😆

mirka commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -15,97 +15,121 @@ import Button from '../../button';
-	const [ confirmVal, setConfirmVal ] = useState(
-		"Hasn't confirmed or cancelled yet"
-	);
+export const VeeeryLongMessage = Template.bind( {} );
+VeeeryLongMessage.args = {
+	children: daText().repeat( 20 ),
+};
```
In our new Storybook strategy we're generally trying to focus more on usefulness as _dev docs_ rather than as _test fixtures_, and this story feels like a very specific fixture for testing overflows. So I think we should remove this, and if it's really worth testing we can move it to a unit test.

chad1008 replied:
Makes sense, story removed ✅

mirka commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -15,97 +15,121 @@ import Button from '../../button';
+	args: {
+		children: daText(),
+	},
 	parameters: {
```
Let's also add this so the code snippets are expanded:

```js
	parameters: {
		docs: { source: { state: 'open' } },
	},
```

(Sorry, we're prepping an exhaustive [guide](https://github.com/WordPress/gutenberg/pull/40153) but it's still a WIP!)

chad1008 replied:
Great suggestion, thanks... and thank you for the link to the guide-in-progress!

mirka commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -15,97 +15,121 @@ import Button from '../../button';
+export const UncontrolledAndWithExplicitOnCancel = Template.bind( {} );
+UncontrolledAndWithExplicitOnCancel.args = {
+	confirmOutput: 'Confirmed!',
+	cancelOutput: 'Cancelled',
```
Mmm I'd probably say this story is also extraneous,  because the `onCancel` prop is pretty self-explanatory and straightforward. It _may_ be worth highlighting if this component had a ton of other callback props, but that's not the case here.

In general I think most `on*` props are not particularly helpful to demonstrate in stories unless there's something unusual about it, or it's hard to discover in the props table for some reason or another. In most cases, just the name + description in the props table should be sufficient.

I'm open to other opinions though. If we do think it's worth having detailed descriptions/examples of stuff like this (e.g. like we do in this component's README.md), we should look into using [MDX](https://storybook.js.org/docs/react/writing-docs/mdx) for our docs so we get more control over the presentation.

chad1008 replied:
Done - removing this also had some knock-on benefits in that we no longer need any logic this story's output compared to the default template.

mirka reviewed (commented):
Nice, thanks for working on this!

The main thing I think we should address is that we can't have uncontrolled modals popping up in Docs view. Unless we suspect that most usages will be uncontrolled, the straightforward way to address this would be to make them all controlled examples by default, like on this [doc page for the Reakit modal](https://reakit.io/docs/dialog/). Another tool we have is the [inline stories option](https://github.com/storybookjs/storybook/blob/next/addons/docs/react/README.md#inline-stories) which isolates each story in an iframe.

Sidenote (cc @ciampo): I played around with type extraction on this one, and (perhaps understandably) the docgen has a problem with these [conditional props](https://github.com/WordPress/gutenberg/blob/5808c70e9acd4a3eaabc5c5328d436eaac4072ff/packages/components/src/confirm-dialog/types.ts#L17-L27). I guess there are a couple of ways around it, but I wonder if this is another reason why we should avoid complicating props conditionals as a matter of API design, like you've been noting recently. (Not saying that this one is particularly complicated, or that API design should be dictated by docgen constraints, but just a thought! Interesting how TypeScript can easily enforce complicated conditions, but how quickly it gets messy when trying to explain the rules in a human-readable format.)

chad1008 committed:
ConfirmDialog stories: set code snippets to open

chad1008 committed:
ConfirmDialog Stories: remove 'Uncontrolled with explicit onCancel' story

chad1008 committed:
ConfrirmDialog Stories: set all stories to be controlled by default (prevents modals popping up on DocsPage

chad1008 committed:
ConfirmDialog: remove initial confirmValue text

chad1008 committed:
ConfirmDialog Stories: remove 'Very Long Message' story

chad1008 committed:
ConfirmDialog Stories: remove 'JSX Message' story

chad1008 committed:
ConfirmDialog Stories: remove daText() knob

chad1008 committed:
ConfirmDialoge Stories: remove JSX control remnants

chad1008 committed:
ConfirmDialog Stories: update Controlled story to use controls and the updated Controlled-by-default template

chad1008 commented:
> Unless we suspect that most usages will be uncontrolled, the straightforward way to address this would be to make them all controlled examples by default, like on this [doc page for the Reakit modal](https://reakit.io/docs/dialog/). Another tool we have is the [inline stories option](https://github.com/storybookjs/storybook/blob/next/addons/docs/react/README.md#inline-stories) which isolates each story in an iframe.

I do think most uses will be Uncontrolled, but looking at the two options, I still think making the examples controlled by default makes the most sense. Getting the positioning of the iframes right feels like more trouble than it's worth (part of why Storybook recommends against that method). Also, due to how the docs pull in the code snippets, the examples remain pretty clean... the extra handler functions we'd add to the template don't even show up, so it works great!

I also removed the initial "Hasn't confirmed yet" text. I think it looks cleaner to start with no text and have `Confirmed!`/`Cancelled` appear when the button is clicked.

One final change: because we've now made the default template Controlled, we can actually use that for the `Controlled` story. That story now uses the main template with a separate code snippet so we have a nice clean example on the DocsPage of that mode.

ciampo commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -15,128 +10,129 @@ import Button from '../../button';
-			<Button variant="primary" onClick={ () => setIsOpen( true ) }>
-				Open ConfirmDialog
-			</Button>
+		<ConfirmDialog
+		isOpen={ isOpen }
+		onConfirm={ handleConfirm }
+		onCancel={ handleCancel }
+		>
+		Would you like to privately publish the post now?
+		</ConfirmDialog>
+		
+		<Heading level={ 1 }>{ confirmVal }</Heading>
+		
+		<Button variant="primary" onClick={ () => setIsOpen( true ) }>
+		Open ConfirmDialog
+		</Button>
```
The indentation on this code snippet looks a bit off (although feel free to ignore in case we decide to remove the `Controlled` story altogether)

```suggestion
			<ConfirmDialog
				isOpen={ isOpen }
				onConfirm={ handleConfirm }
				onCancel={ handleCancel }
			>
				Would you like to privately publish the post now?
			</ConfirmDialog>
	
			<Heading level={ 1 }>{ confirmVal }</Heading>
		
			<Button variant="primary" onClick={ () => setIsOpen( true ) }>
				Open ConfirmDialog
			</Button>
```

chad1008 replied:
good catch, I didn't spot that in the template literal. I've fixed it, and also replaced the tabs with double spaces... it looks like Storybook is using double space indentation when it autogenerates from the template, so my tabs looked off when a manual snippet was rendered just above an auto generated one. This way they all match up in terms of formatting.

ciampo commented:
Thank you @chad1008 for working on this!

> Unless we suspect that most usages will be uncontrolled, the straightforward way to address this would be to make them all controlled examples by default

Agree with @mirka on this one

> One final change: because we've now made the default template Controlled, we can actually use that for the Controlled story. That story now uses the main template with a separate code snippet so we have a nice clean example on the DocsPage of that mode.

If I understand correctly, since currently all examples are using the controlled version of the component, I don't think we need a "Controlled" Story anymore?

At most, we'd need a way to flag that the component can also be used in _uncontrolled_ mode — correct?

> I wonder if this is another reason why we should avoid complicating props conditionals as a matter of API design

You raise a good point — I've never been a fan of conditional APIs, and I don't really like the solution that we adopted on `ConfirmDialog` (which I think I suggested in the first place!) — I wonder if we should just export two separate components instead?

(Coincidentally, the same pattern is being discussed in [the `BoxModelOverlay` PR](https://github.com/WordPress/gutenberg/pull/40253#discussion_r852179379))

---

Finally, I have a couple of questions / observations:

- since the rest of the component is written in TypeScript, should we write these stories in TypeScript too ? (it's also ok if we want to do that in a follow-up PR!)
- We should probably add JSDoc comments to the props defined in [`types.ts`](https://github.com/WordPress/gutenberg/blob/aca7e6ee495cf6cbbab283ab682c1b81c9070797/packages/components/src/confirm-dialog/types.ts), so that they would displayed in the Storybook auto-generated docs

chad1008 committed:
replace the Controlled story with Uncontrolled now that Controlled is the default

chad1008 commented:
> If I understand correctly, since currently all examples are using the controlled version of the component, I don't think we need a "Controlled" Story anymore?
> At most, we'd need a way to flag that the component can also be used in uncontrolled mode — correct?

You make a good point. I've reworked it a bit so that the default example now includes the full Controlled snippet, with the state management callbacks shown. I've then replaced the separate Controlled story with a simple Uncontrolled story instead, to demo a basic application without needing to declare `isOpen` and `onCancel`.

> since the rest of the component is written in TypeScript, should we write these stories in TypeScript too ? (it's also ok if we want to do that in a follow-up PR!)

Absolutely, I definitely plan on addressing this in a follow-up!

> We should probably add JSDoc comments to the props defined in [types.ts](https://github.com/WordPress/gutenberg/blob/aca7e6ee495cf6cbbab283ab682c1b81c9070797/packages/components/src/confirm-dialog/types.ts), so that they would displayed in the Storybook auto-generated docs

I had the same thought, but found that the JSDoc comments I added there didn't get added to the the the auto-generated docs. I initially wondered if this was because the stories weren't in TypeScript yet, but I'm no wondering if it's a side effect of those conditional props? Are they possibly preventing Storybook from parsing them the way the would on static props?

ciampo commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -121,22 +56,103 @@ export const Controlled = () => {
+	confirmButtonText: 'Yes please!',
+};
+
+const uncontrolledSnippet = `<>
+  <ConfirmDialog
+    onConfirm={ handleConfirm }
+  >
+    Would you like to privately publish the post now?
+  </ConfirmDialog>
+  
+  <Heading level={ 1 } />
+</>
+`;
+
+export const Uncontrolled = Template.bind( {} );
+Uncontrolled.args = {};
+Uncontrolled.parameters = {
+	docs: {
+		description: {
+			story:
+				'To render in Uncontrolled Mode, omit passing a boolean to the `isOpen` prop. This will allow the component to close itself without the need for an explicit callback. In Uncontrolled mode, `onCancel` is optional.',
+		},
+		source: {
+			code: uncontrolledSnippet,
+			language: 'jsx',
+			type: 'auto',
+		},
+	},
+};
```
Looking at this, I'm not actually sure if we should have an uncontrolled example:

- this uncontrolled example wouldn't really add any value to the  would be probably misleading (it uses the "controlled" version of the component to showcase the uncontrolled?)
- the snippet, in its current status, is somewhat incomplete (`handleConfirm` is not defined anywhere, the `Heading` component doesn't have text and is not relevant to the example)

I still think that we should somehow document the uncontrolled mode, though. Currently, [we do mention it in the README](https://github.com/WordPress/gutenberg/blob/aca7e6ee495cf6cbbab283ab682c1b81c9070797/packages/components/src/confirm-dialog/README.md#uncontrolled-mode), but we should consider how we can do that effectively in Storybook if we eventually want to move all of our docs.

@mirka what do you think?

mirka replied:
Given that the conditional props will [prevent](https://github.com/WordPress/gutenberg/pull/40164#issuecomment-1111466890) us from getting a clean props table anyway, it might be worth considering splitting it into separate components (e.g. `ConfirmDialog` and `ConfirmDialog.Uncontrolled`?), or maybe even removing the uncontrolled version.

ciampo replied:
That's an interesting point.

For the sake of this PR, I guess we could remove the uncontrolled example completely from this PR, and tackle the conditional props and the uncontrolled component in a separate PR

chad1008 replied:
I've just looked over the four current implementations of `ConfirmDialog` in Gutenberg, and currently they're all Controlled, which lends more weight to the idea of doing away with Uncontrolled altogether.

I agree with removing the Uncontrolled story for now at the very least, and we can perhaps update the component to only have the single Controlled mode in a followup - which would also eliminate the conditional props at the same time which would be a double-win.

ciampo commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -121,22 +56,103 @@ export const Controlled = () => {
+Uncontrolled.parameters = {
+	docs: {
+		description: {
+			story:
+				'To render in Uncontrolled Mode, omit passing a boolean to the `isOpen` prop. This will allow the component to close itself without the need for an explicit callback. In Uncontrolled mode, `onCancel` is optional.',
+		},
```
I verified on my local machine that in this case it wasn't possible to assign the story description as a comment — it must be because we're explicitly passing `parameters.docs` in this example? (cc @mirka )

mirka replied:
Exactly! I have a fix for this in #40616.

ciampo replied:
Fantastic! I'm going to mark this convo as resolved

ciampo commented on packages/components/src/confirm-dialog/stories/index.js:
```diff
@@ -121,22 +56,103 @@ export const Controlled = () => {
+			</Button>
+
 			<ConfirmDialog
 				isOpen={ isOpen }
 				onConfirm={ handleConfirm }
 				onCancel={ handleCancel }
+				cancelButtonText={ args.cancelButtonText }
+				confirmButtonText={ args.confirmButtonText }
 			>
-				{ daText() }
+				{ args.children }
 			</ConfirmDialog>
```
Nit: we can simplify the code here:

```jsx
			<ConfirmDialog
				{ ...args }
				isOpen={ isOpen }
				onConfirm={ handleConfirm }
				onCancel={ handleCancel }
			/>
```

chad1008 replied:
Good catch! Updated ✅

ciampo reviewed (commented):
> I had the same thought, but found that the JSDoc comments I added there didn't get added to the the the auto-generated docs. I initially wondered if this was because the stories weren't in TypeScript yet, but I'm no wondering if it's a side effect of those conditional props? Are they possibly preventing Storybook from parsing them the way the would on static props?

I believe that the main reason why types are not coming through is because the stories are not written in TypeScript yet — apologies if I didn't explain that more clearly in my previous message.

> Absolutely, I definitely plan on addressing this in a follow-up!

That would be great, thank you!

mirka commented:
> I had the same thought, but found that the JSDoc comments I added there didn't get added to the the the auto-generated docs. I initially wondered if this was because the stories weren't in TypeScript yet, but I'm no wondering if it's a side effect of those conditional props?

Yes, those conditional props are the main blocker at the moment, which I don't mind punting to a separate PR. There's also another prerequisite for the type extraction to work, which is to add a named export [like so](https://github.com/WordPress/gutenberg/blob/411547efb48744c0413152cb1a477f5b393518cf/packages/components/src/divider/component.tsx#L56). But in any case the docgen part won't work unless we do something about the conditional props.

ciampo commented:
> > I had the same thought, but found that the JSDoc comments I added there didn't get added to the the the auto-generated docs. I initially wondered if this was because the stories weren't in TypeScript yet, but I'm no wondering if it's a side effect of those conditional props?
> 
> Yes, those conditional props are the main blocker at the moment, which I don't mind punting to a separate PR. There's also another prerequisite for the type extraction to work, which is to add a named export [like so](https://github.com/WordPress/gutenberg/blob/411547efb48744c0413152cb1a477f5b393518cf/packages/components/src/divider/component.tsx#L56). But in any case the docgen part won't work unless we do something about the conditional props.

Let's do that in a separate PR, so that in the meantime we can wrap up the refactor to controls.

Next follow-ups will be:

- figure out what to do with the uncontrolled version of the component (and therefore simplify the conditional props)
- after that, we'll be able to convert the Storybook examples to TypeScript and finally be able to generate docs

WDTY ?

chad1008 committed:
ConfirmDialog Stories: remove Uncontrolled Story

chad1008 committed:
ConfirmDialog Stories: simplify args in template code

chad1008 commented:
I've just pushed a couple more updates cleaning up the args in the template as @ciampo suggested and removing the Uncontrolled story for the time being.

Knowing that all existing uses of the component are currently Controlled, I'm leaning more and more in favor of removing Uncontrolled as an option completely in a followup PR.

Once we do that, the TypeScript conversion of these stories (another followup) should be the main thing needed to wrap this up completely!

mirka reviewed (approved):
> Knowing that all existing uses of the component are currently Controlled, I'm leaning more and more in favor of removing Uncontrolled as an option completely in a followup PR.
> 
> Once we do that, the TypeScript conversion of these stories (another followup) should be the main thing needed to wrap this up completely!

Sounds good! Thanks for all your work on this 🚀
